netket/models/equivariant.py

```diff
@@ -24,54 +24,271 @@
+    """Array specifying the characters of the desired symmetry representation"""
+    dtype: Any = float
+    """The dtype of the weights."""
+    activation: Any = jax.nn.selu
+    """The nonlinear activation function between hidden layers."""
+    output_activation: Any = identity
+    """The nonlinear activation before the output. Defaults to the identity."""
+    imag_part: bool = False
+    """If true return only the imaginary part of the output"""
+    use_bias: bool = True
+    """if True uses a bias in all layers."""
+    precision: Any = None
+    """numerical precision of the computation see `jax.lax.Precision`for details."""
+    kernel_init: NNInitFunc = unit_normal_scaling
+    """Initializer for the Dense layer matrix."""
+    bias_init: NNInitFunc = zeros
+    """Initializer for the hidden bias."""
+
+    def setup(self):
+
+        self.n_symm = np.asarray(self.symmetries).shape[0]
+
+        self.dense_symm = DenseSymmFFT(
+            space_group=self.symmetries,
+            shape=self.shape,
+            features=self.features[0],
+            dtype=self.dtype,
+            use_bias=self.use_bias,
+            kernel_init=self.kernel_init,
+            bias_init=self.bias_init,
+            precision=self.precision,
+        )
+
+        self.equivariant_layers = [
+            DenseEquivariantFFT(
+                product_table=self.product_table,
+                shape=self.shape,
+                in_features=self.features[layer],
+                out_features=self.features[layer + 1],
+                use_bias=self.use_bias,
+                dtype=self.dtype,
+                precision=self.precision,
+                kernel_init=self.kernel_init,
+                bias_init=self.bias_init,
+            )
+            for layer in range(self.layers - 1)
+        ]
+
+    @nn.compact
+    def __call__(self, x):
+        x = self.dense_symm(x)
+
+        for layer in range(self.layers - 1):
+            x = self.activation(x)
+            x = self.equivariant_layers[layer](x)
+
+        x = self.output_activation(x)
+
+        x = x.reshape(-1, self.features[-1] * self.n_symm)
+        x_max = jnp.max(x, axis=-1, keepdims=True)
+        x = jnp.exp(x - x_max)
+        x = x.reshape(-1, self.features[-1], self.n_symm)
+        x = jnp.sum(x, 1)
+
+        x = jnp.sum(x * jnp.array(self.characters), -1)
+
+        x = jnp.log(x) + jnp.squeeze(x_max)
+
+        if self.imag_part:
+            return 1j * jnp.imag(x)
+        else:
+            return x
 
-    symmetries: Union[HashableArray, PermutationGroup]
-    """A group of symmetry operations (or array of permutation indices) over which the layer should be invariant.
 
+class GCNN_Irrep(nn.Module):
+    """Implements a GCNN by projecting onto irreducible
+    representations of the group. The projection onto
+    the group is implemented with matrix multiplication"""
+
+    """Layers act on a feature maps of shape [batch_size, in_features, n_symm] and 
```

Comment:
```suggestion
    the group is implemented with matrix multiplication

    Layers act on a feature maps of shape [batch_size, in_features, n_symm] and 
```

you need a newline, not to make two different strings.